Feat/8763 periodic settlement data endpoint

guoguojin opened the pull request:
Closes #8763

guoguojin committed:
chore: generate protos and add db migration scripts

candida-d commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -1943,6 +1943,14 @@ type Market {
   "Optional: Market ID of the successor to this market if one exists"
   successorMarketID: ID
+
+  "Funding Periods for perpetual markets"
```
```suggestion
  "Funding periods for perpetual markets"
```

edd replied:
I agree - this would be good to have in the field description. Based on https://github.com/vegaprotocol/vega/blob/97980ebf5ba285076bf3a5eb5a4e0c23706df08f/datanode/gateway/graphql/schema.graphql#L287-L288 I guess it's a RFC3339Nano?

candida-d commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -5188,3 +5196,82 @@ type Entities {
+  marketId: ID!
+  # Sequence number of the funding period.
+  seq: Int!
+  # Time when the funding period started.
```
how is this time measured/expressed?

candida-d commented on protos/sources/data-node/api/v2/trading_data.proto:
```diff
@@ -713,6 +713,20 @@ service TradingDataService {
 
+  // List funding periods
+  //
+  // Get a list of funding periods for a market
```
```suggestion
  // Get a list of funding periods for a perpetual market
```

candida-d commented on protos/sources/data-node/api/v2/trading_data.proto:
```diff
@@ -713,6 +713,20 @@ service TradingDataService {
+
+  // List funding period data points
+  //
+  // Get a list of data points for a market's funding periods
```
```suggestion
  // Get a list of data points for a perpetual market's funding periods
```

candida-d commented on protos/sources/data-node/api/v2/trading_data.proto:
```diff
@@ -3312,6 +3326,72 @@ message ListEntitiesResponse {
   repeated vega.events.v1.ProtocolUpgradeEvent protocol_upgrade_proposals = 28;
 }
 
+// Request to list a market's funding periods
```
```suggestion
// Request to list a perpetual market's funding periods
```

candida-d commented on protos/sources/data-node/api/v2/trading_data.proto:
```diff
@@ -3312,6 +3326,72 @@ message ListEntitiesResponse {
+message ListFundingPeriodsRequest {
+  // Market ID to get funding periods for.
+  string market_id = 1 [(google.api.field_behavior) = REQUIRED];
+  // Optional funding period sequence number to limit results to a specific funding period.
```
would it be clear to someone using these APIs how to get the funding period sequence number? (This is a question for all the instances that mention this sequence number)

guoguojin replied:
I don't think so, but @edd has asked for a date range as input which makes much more sense so I'm just going to remove this option in the request

witgaw commented on datanode/entities/fund_period.go:
```diff
@@ -0,0 +1,202 @@
+	"code.vegaprotocol.io/vega/libs/num"
+)
+
+type FundingPeriod struct {
```
This should also include:
1. Market price TWAP (time-weighted average price coming from the perps market on Vega
2. Reference price TWAP (time-weighed average price coming from the oracle)

guoguojin replied:
This isn't being provided by core, only the data points. The question is should the TWAP market/reference price be provided by core, or do you want this to be calculated from the data points by data node? The entity here represents the database object that we map from the core event we receive so this may or may not be updated depending on the answer.

guoguojin replied:
Confirmed with @witgaw that the TWAP information should be emitted from core.

witgaw commented on datanode/entities/fund_period.go:
```diff
@@ -0,0 +1,202 @@
+package entities
```
Could you rename the file to `funding_period.go`?

We might have fund accounts eventually and they might have some periods, so this might end up being confusing.

guoguojin replied:
Yes, sorry, that was a typo, I will fix.

edd commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -1943,6 +1943,14 @@ type Market {
   successorMarketID: ID
+
+  "Funding Periods for perpetual markets"
+  fundingPeriods(
```
**question (blocking)**: Checking other platforms, it looks like funding periods typically take a date parameter or two:

(dydx v4 uses `.../v4/historicalFunding/BTC-USD?effectiveBeforeOrAt=2023-07-06T21:59:59.899Z&null=2023-07-02T18:59:59.899Z`, okx `.../funding-rate-history-new?t=1690480575735&ccy=BTC`) . Is there a reason we don't provide that? 

Presumably this would be used for charting the funding rates. Date controls would help there.

**Basic suggestion**: An optional new parameter, `endDate`, that can be used to fetch `n` funding periods up until a dateTime (where `n` is defined by pagination or a sensible default)

**Fancier suggestion**: Two optional parameters, `startDate` and `endDate`.

Is there arny reason we can't get those?

guoguojin replied:
There's no reason we can't add it, happy to do so.

edd reviewed (changes requested):
Feature request: date/date range for fundingPeriods

guoguojin committed:
chore: periodic settlement data endpoint for perpetuals

guoguojin committed:
chore: update proto and gql documentation and regenerate

guoguojin committed:
chore: address code review suggestions

- Add external and internal TWAP values for funding period and data
  points.
- Update API to use a date range rather than the funding period funding
  period.

guoguojin commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -1943,6 +1943,14 @@ type Market {
   successorMarketID: ID
+
+  "Funding periods for perpetual markets"
+  fundingPeriods(
```
@mattrussell36 do you want the funding periods data points API which I nested as a sub-query under funding periods to be a separate top level query too, as that also allows for pagination of the data.

guoguojin committed:
chore: make funding periods a top level GQL query

guoguojin committed:
chore: fix API documentation

witgaw reviewed (dismissed):
LGTM

mattrussell36 commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -5188,3 +5222,79 @@ type Entities {
+  "RFC3339Nano time when the funding period started."
+  startTime: Timestamp!
+  "RFC3339Nano time when the funding period ended."
+  endTime: Timestamp
```
@EVODelavega @jeremyletang 

At what point is a funding event created? And at what point is the endTime value set?

We need to be sure UIs can show when the current funding period will end (or the next one will start). Either by using the endTime field here, or by using the startTime of the next period.

mattrussell36 replied:
I've spoken with @jeremyletang in person and I think we are good, we'll be using a time trigger from the products data source, adding the interval duration to the current funding periods startTime

edd reviewed (dismissed):
Thanks for the date ranges 👍
